compiler-core/src/analyse.rs

```diff
@@ -247,7 +247,40 @@ impl<'a, A> ModuleAnalyzer<'a, A> {
             self.register_type_alias(t, &mut env);
         }
 
+        for c in &statements.constants {
+            // don't warn if the definition is a re-export.
+            if let Constant::<(), ()>::Var {
+                location: _,
+                module: _,
+                name,
+                constructor: _,
+                type_: _,
+            } = &*c.value
+            {
+                if name == &c.name {
+                    continue;
+                }
+            }
+
+            if env.unqualified_imported_names.contains_key(&c.name) {
+                self.problems
+                    .warning(Warning::TopLevelDefinitionShadowsImport {
+                        location: c.location,
+                        name: c.name.clone(),
+                    })
+            }
+        }
+
         for f in &statements.functions {
+            let (_name_location, name) = f.name.clone().expect("A module's function must be named");
+            if env.unqualified_imported_names.contains_key(&name) {
+                self.problems
+                    .warning(Warning::TopLevelDefinitionShadowsImport {
+                        location: f.location,
+                        name,
+                    });
+            }
```

Comment:
Please move the code for this check into the existing definition analysis rather than iterating over the functions and constants again 🙏